### PR TITLE
Pretimodit

### PR DESCRIPTION
**Pull request by HajimeKawahara**
Draft PR for new algorithm for PreMODIT.

## Elower max
See `tests/unittests/spec/optgrid_test.py`

We made `elower_max` option in MdbExomol and MdbHitemp

#331

## New PreMODIT algorithm

I confirmed that the new algorithm works well for ExoMol, but some small errors for Hitemp may be due to n_Texp or something.
 
See examples using the current implementation in 
`test/unittests/spec/premodit/premodit_spectrum_test.py`
`test/unittests/spec/premodit/premodit_xsection_test.py`

- diffmode = 0, PreMODIT in current develop branch
- diffmode = 1, the use of 1st derivative of weight
- diffmode = 2, the use of 1st and 2nd derivatives of weight

For retrieval, try
`tests/endtoend/reverse/reverse_premodit.py`

#317
### accuracy
You can check the accuracy as a function of parameters in elower_grid.ipynb

## OpaPremodit class

easy to call premodit... Now we have two important classes, one is `mdb` and another is `opa`.
`opa`  (`OpaPremodit` for Premodit) sets control parameters for opacity calculation. 
Then, you can directly call the cross section computation, `opa.xsmatrix` (for multiple T and P) or `opa.xsvector` (for single T and P):

```python
from exojax.spec.api import MdbExomol
from exojax.spec.opacalc import OpaPremodit

mdb = MdbExomol('.database/CH4/12C-1H4/YT10to10/',
                nurange=nu_grid,
                gpu_transfer=False)

opa = OpaPremodit(mdb=mdb,
                  nu_grid=nu_grid,
                  diffmode=diffmode,
                  auto_trange=[400.0, 1500.0])
xsm = opa.xsmatrix(Tarr, Parr)

```

## Documents

- [Forward Modeling using PreMODIT](http://secondearths.sakura.ne.jp/exojax/develop/tutorials/Forward_modeling_using_PreMODIT_Cross_Section_for_methane.html)

## I will make 

- [x] auto parameters setting based on Tmax, Tmin, accuracy = lbderror/optimal_params
- [x] auto params in initspec 
- [x] tests for reverse modling
- [x] bug fix Hitemp
- [x] document

### CODE REVIEW

**Comment by HajimeKawahara** [comment]
```
        with open(filename, 'rb') as f:
>           mdb = pickle.load(f)
E           ModuleNotFoundError: No module named 'vaex.hdf5'
```
something (maybe version mismatch) wrong for vaex...

**Comment by HajimeKawahara** [comment]
Note that the auto unit tests failed due to vaex issue (all of the unit tests passed in my environment)
```
E       ModuleNotFoundError: No module named 'vaex.hdf5'
```

**Review by YuiKasagi — APPROVED** [review]
I tried using OpaPremodit with my optimization (jaxopt) code and it worked. Thank you for the update.

**Comment by HajimeKawahara** [comment]
@YuiKasagi Thanks! Did you need to restrict the temperature range?, like
```python
Tarr = jnp.clip(T0_in * Parr**alpha_in, Tlow, Thigh)
```

**Comment by YuiKasagi** [comment]
@HajimeKawahara No, I did not restrict it. I set ```Tarr = T0*(Parr/Pref)**alpha``` and worked for H2O fitting for my target.

File: setup.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 ]
 INSTALL_REQUIRES = [
     'numpy<=1.22.3', 'pandas', 'tqdm', 'scipy', 'jax>=0.2.22', 'numpyro',
-    'pyarrow', 'matplotlib <= 3.6.0', 'hitran-api', 'bs4', 'vaex>=4.9.2',
+    'pyarrow', 'matplotlib <= 3.6.0', 'hitran-api', 'bs4', 'vaex', 'vaex-hdf5',
```
Comment (erwanp) [inline setup.py]: Also see recent remark on Numpy constraint which might not be required anymore https://github.com/radis/radis/issues/548

Comment (HajimeKawahara) [reply setup.py]: using pip, constraint in setup.py of radis prevents dependency check. Should we update setup.py in radis first?

```
Installed /opt/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/PeakUtils-1.3.4-py3.8.egg
error: numpy 1.24.2 is installed but numpy<=1.22.3 is required by {'radis'}
Error: Process completed with exit code 1.
```

**Comment by chonma0ctopus** [comment]
Thank you for the development. 
Sorry for the late review but I confirmed that the tests and examples work also in my environment.
However, I found that `Tref` becomes `None` during the optimization process if I set the upper T (`Tu`) in `auto_trange` to 5000 K for securing the precision for the deep atmosphere of M dwarfs.

For example,
```
from exojax.utils.grids import wavenumber_grid
from exojax.spec import api
from exojax.spec.opacalc import OpaPremodit

from jax.config import config
config.update("jax_enable_x64", True)

nu_grid, wav, res = wavenumber_grid(10874., 10916., 3000, unit='AA', xsmode='premodit')
mdbH2O = api.MdbExomol('.database/H2O/1H2-16O/POKAZATEL', nu_grid, gpu_transfer=False)
opa = OpaPremodit(mdb=mdbH2O, nu_grid=nu_grid, diffmode=1, 
                     auto_trange=[1000.0, 5000.0], dit_grid_resolution=0.2)
# (It works for auto_trange=[1000.0, 4900.0] but not for auto_trange=[1000.0, 5000.0].)
```

returns the error below:
```
xsmode =  premodit
xsmode assumes ESLOG in wavenumber space: mode=premodit
HITRAN exact name= H2(16O)
Background atmosphere:  H2
Reading .database/H2O/1H2-16O/POKAZATEL/1H2-16O__POKAZATEL__09100-09200.trans.bz2
.broad is used.
Broadening code level= a1
default broadening parameters are used for  10  J lower states in  61  states
OpaPremodit: params automatically set.
/home/tako/ghR/exojax/src/exojax/spec/lbderror.py:294: UserWarning: Couldn't find the params.
  warnings.warn("Couldn't find the params.")
Tref changed: 296.0K->NoneK
Traceback (most recent call last):
  File "/home/tako/work/bugreport_OpaPremodit_MdwarfsT.py", line 6, in <module>
    opa = OpaPremodit(mdb=mdbH2O, nu_grid=nu_grid, diffmode=1, 
  File "/home/tako/ghR/exojax/src/exojax/spec/opacalc.py", line 71, in __init__
    self.auto_setting(auto_trange[0], auto_trange[1])
  File "/home/tako/ghR/exojax/src/exojax/spec/opacalc.py", line 83, in auto_setting
    self.apply_params()
  File "/home/tako/ghR/exojax/src/exojax/spec/opacalc.py", line 118, in apply_params
    self.mdb.change_reference_temperature(self.Tref)
  File "/home/tako/ghR/exojax/src/exojax/spec/api.py", line 288, in change_reference_temperature
    qr = self.qr_interp(Tref_new)
  File "/home/tako/ghR/exojax/src/exojax/spec/api.py", line 279, in qr_interp
    return self.QT_interp(T) / self.QT_interp(self.Tref)
  File "/home/tako/ghR/exojax/src/exojax/spec/api.py", line 268, in QT_interp
    return jnp.interp(T, self.T_gQT, self.gQT)
  File "/home/tako/miniconda3/envs/base+newJAX/lib/python3.9/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/tako/miniconda3/envs/base+newJAX/lib/python3.9/site-packages/jax/_src/api.py", line 620, in cache_miss
    execute = dispatch._xla_call_impl_lazy(fun_, *tracers, **params)
  File "/home/tako/miniconda3/envs/base+newJAX/lib/python3.9/site-packages/jax/_src/dispatch.py", line 236, in _xla_call_impl_lazy
    return xla_callable(fun, device, backend, name, donated_invars, keep_unused,
  File "/home/tako/miniconda3/envs/base+newJAX/lib/python3.9/site-packages/jax/linear_util.py", line 300, in memoized_fun
    ans = call(fun, *args)
  File "/home/tako/miniconda3/envs/base+newJAX/lib/python3.9/site-packages/jax/_src/dispatch.py", line 359, in _xla_callable_uncached
    return lower_xla_callable(fun, device, backend, name, donated_invars, False,
  File "/home/tako/miniconda3/envs/base+newJAX/lib/python3.9/site-packages/jax/_src/profiler.py", line 314, in wrapper
    return func(*args, **kwargs)
  File "/home/tako/miniconda3/envs/base+newJAX/lib/python3.9/site-packages/jax/_src/dispatch.py", line 445, in lower_xla_callable
    jaxpr, out_type, consts = pe.trace_to_jaxpr_final2(
  File "/home/tako/miniconda3/envs/base+newJAX/lib/python3.9/site-packages/jax/_src/profiler.py", line 314, in wrapper
    return func(*args, **kwargs)
  File "/home/tako/miniconda3/envs/base+newJAX/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 2081, in trace_to_jaxpr_final2
    jaxpr, out_type, consts = trace_to_subjaxpr_dynamic2(fun, main, debug_info)
  File "/home/tako/miniconda3/envs/base+newJAX/lib/python3.9/site-packages/jax/interpreters/partial_eval.py", line 2031, in trace_to_subjaxpr_dynamic2
    ans = fun.call_wrapped(*in_tracers_)
  File "/home/tako/miniconda3/envs/base+newJAX/lib/python3.9/site-packages/jax/linear_util.py", line 167, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/tako/miniconda3/envs/base+newJAX/lib/python3.9/site-packages/jax/_src/numpy/lax_numpy.py", line 1000, in interp
    _check_arraylike("interp", x, xp, fp)
  File "/home/tako/miniconda3/envs/base+newJAX/lib/python3.9/site-packages/jax/_src/numpy/util.py", line 345, in _check_arraylike
    raise TypeError(msg.format(fun_name, type(arg), pos))
jax._src.traceback_util.UnfilteredStackTrace: TypeError: interp requires ndarray or scalar arguments, got <class 'NoneType'> at position 0.

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/tako/work/bugreport_OpaPremodit_MdwarfsT.py", line 6, in <module>
    opa = OpaPremodit(mdb=mdbH2O, nu_grid=nu_grid, diffmode=1, 
  File "/home/tako/ghR/exojax/src/exojax/spec/opacalc.py", line 71, in __init__
    self.auto_setting(auto_trange[0], auto_trange[1])
  File "/home/tako/ghR/exojax/src/exojax/spec/opacalc.py", line 83, in auto_setting
    self.apply_params()
  File "/home/tako/ghR/exojax/src/exojax/spec/opacalc.py", line 118, in apply_params
    self.mdb.change_reference_temperature(self.Tref)
  File "/home/tako/ghR/exojax/src/exojax/spec/api.py", line 288, in change_reference_temperature
    qr = self.qr_interp(Tref_new)
  File "/home/tako/ghR/exojax/src/exojax/spec/api.py", line 279, in qr_interp
    return self.QT_interp(T) / self.QT_interp(self.Tref)
  File "/home/tako/ghR/exojax/src/exojax/spec/api.py", line 268, in QT_interp
    return jnp.interp(T, self.T_gQT, self.gQT)
  File "/home/tako/miniconda3/envs/base+newJAX/lib/python3.9/site-packages/jax/_src/numpy/lax_numpy.py", line 1000, in interp
    _check_arraylike("interp", x, xp, fp)
  File "/home/tako/miniconda3/envs/base+newJAX/lib/python3.9/site-packages/jax/_src/numpy/util.py", line 345, in _check_arraylike
    raise TypeError(msg.format(fun_name, type(arg), pos))
TypeError: interp requires ndarray or scalar arguments, got <class 'NoneType'> at position 0.
```

Could you have some suggestions?

**Comment by HajimeKawahara** [comment]
@chonma0ctopus Thanks for the report. This is because I assumed <5000K for premodit auto_trange. We should rethink this issue, i.e. the temperature range. I will make new issue for this.

**Pull request merged**